### ISSUE TITLE
Guidebot Looks for Reactor First

### ISSUE THREAD

**Issue by Mako88** [dxx-rebirth/dxx-rebirth#271]
In D2, guidebot starts looking for the reactor instead of the keys first.

I bisected it and got 03b022289625954b6b4c5a370308769c7d99819c as the bad commit.

Also, I was getting an intermittent crash after telling guidebot to search for various items (specifically hostages). I'm honestly not certain what exact steps resulted in the crash, other than telling guidebot to search for hostages on D2 Counterstrike level 1. I wasn't able to get a backtrace (yet), but I did get this:

```
terminate called after throwing an instance of 'valptridx::index_range_exception'what():  d2x-rebirth/main/escort.cpp:331: invalid index used in array subscript: base=0xd71bc0 size=350 index=65534
```

@zicodxx thought this may be caused by the same thing since this is an index out of bounds error with the guidebot's goals.

**Comment by Mako88** [comment]
And here's a backtrace:

```
#0  0x00007ffff63a204f in raise () from /usr/lib/libc.so.6
#1  0x00007ffff63a347a in abort () from /usr/lib/libc.so.6
#2  0x00007ffff6cb94bd in __gnu_cxx::__verbose_terminate_handler ()
    at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007ffff6cb7276 in __cxxabiv1::__terminate (handler=<optimized out>)
    at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007ffff6cb72c1 in std::terminate ()
    at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00007ffff6cb74d8 in __cxxabiv1::__cxa_throw (obj=obj@entry=0x181e4ad0, 
    tinfo=tinfo@entry=0x506038 <typeinfo for valptridx<d2x::object>::index_range_exception>, 
    dest=dest@entry=0x47d480 <valptridx<d2x::object>::index_range_exception::~index_range_exception()>)
    at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc:87
#6  0x0000000000407f23 in valptridx<d2x::object>::index_range_exception::report
    (filename=filename@entry=0x518c5d "d2x-rebirth/main/escort.cpp", 
    lineno=lineno@entry=331, array=<optimized out>, 
    supplied_index=<optimized out>) at common/main/valptridx.tcc:95
#7  0x00000000004e63fe in valptridx<d2x::object>::check_index_range<std::less>
    (a=<optimized out>, i=<optimized out>, lineno=331, 
    filename=<optimized out>) at common/include/valptridx.h:128
#8  valptridx<d2x::object>::basic_ptr<valptridx<d2x::object>::vc, 0u>::basic_ptr (a=..., i=<optimized out>, lineno=331, filename=<optimized out>, 
---Type <return> to continue, or q <return> to quit---
    this=<optimized out>) at common/include/valptridx.h:387
#9  valptridx<d2x::object>::basic_vptr_global_factory<valptridx<d2x::object>::basic_ptr<valptridx<d2x::object>::vc, 0u> >::operator() (lineno=331, 
    filename=<optimized out>, i=<optimized out>, this=<synthetic pointer>)
    at common/include/valptridx.h:671
#10 d2x::detect_escort_goal_accomplished (index=...)
    at d2x-rebirth/main/escort.cpp:331
#11 0x0000000000437d0c in d2x::collide_robot_and_weapon (robot=..., 
    weapon=..., collision_point=...) at similar/main/collide.cpp:1747
#12 0x000000000043894c in collide_two_objects (A=..., B=..., 
    collision_point=...) at similar/main/collide.cpp:2526
#13 0x00000000004aa8c8 in d2x::do_physics_sim (obj=...)
    at similar/main/physics.cpp:653
#14 0x00000000004a8b53 in d2x::object_move_one (obj=...)
    at similar/main/object.cpp:1736
#15 d2x::object_move_all () at similar/main/object.cpp:1882
#16 0x000000000044c575 in d2x::GameProcessFrame ()
    at similar/main/game.cpp:1504
#17 d2x::game_handler (event=...) at similar/main/game.cpp:1272
#18 0x0000000000410028 in dcx::window_send_event (event=..., wind=...)
    at common/include/window.h:116
#19 dcx::event_process () at common/arch/sdl/event.cpp:169
#20 0x000000000040aa96 in d2x::main (
---Type <return> to continue, or q <return> to quit---
    argv_gd$MC41OC4xLWQyeC03MzEwLWc4ODE3MzA3ZjIzZDAĽ=<optimized out>, 
    argc_gc$8817307f23d0f995d7f14366d280f13e7f24d7a4=<optimized out>)
    at similar/main/inferno.cpp:596
#21 main (argc_gc$8817307f23d0f995d7f14366d280f13e7f24d7a4=<optimized out>, 
    argv_gd$MC41OC4xLWQyeC03MzEwLWc4ODE3MzA3ZjIzZDAĽ=<optimized out>)
    at similar/main/inferno.cpp:634
```

**Comment by zicodxx** [comment]
I first suspected the crash would be related to the commit but now I am not so certain anymore. But the commit seems logical to be the culprit:
Descent 2 - Level 1 does not have a Blue now Yellow key. The approach to reduce the calls of exists_in_mine() is nice but this way, the code will find that the player does not own the blue key, exists_in_mine() will say that it doesn't exist either and jump straight to the check for reactor/boss where it finds the former. So checking for a possible existence of a Yellow or Red key, isn't even considered.

I would actually revert this commit since I see no way how to solve this without invoking multiple calls if exists_in_mine() at least not in a more efficient way as it was before the commit. 
As far as I know, there is no ruleset of which keys need to be present. There could be any combination of keys present in the level.

**Comment by vLKp** [comment]
Yes, that commit was a mistake.  That type of transform is only correct on a similar layout that is not present here.

The crash on invalid index is because something received `object_guidebot_cannot_reach` and decided to treat it as a valid index, which it is not.

**Comment by vLKp** [comment]
The escort goal handling is a mess.  Various indexes go into different ambiguously named globals.  Sometimes those are indexes of objects.  Other times they are indexes of segments.

**Comment by zicodxx** [comment]
I suppose the Escort_goal_index could always point to a segment as every potential escort object is located in one. But I don't like that since segments can be huge and the Guidebot would stop its goal soon as reaching a potentially massive area.

Instead, how about we take Escort_goal_index (either an object or segment), Escort_goal_object and Escort_special_goal (both guidebot goal list entries) and stuff them into an Escort_goal structure?
(Escort_)goal and (Escort_)special_goal could more or less stay as they are - tho from naming, I would rather call them "(special)_index".

Escort_goal_index would then be split in goal_obj and escort_seg. While looking for an object, goal_obj points to that and escort_seg points to obj->segnum. If the goal is a segment, escort_obj could point to object_none. 

While setting the actual goal, we may also have a variable to pinpoint specifically if the goal is an object or a segment (could even be a goal: goal_is_object - if false, it's assumed a segment). This could be determined from the actual goal (i.e. current Escort_goal_object). But I suppose we don't even need that if we always maintain a valid object and segment for the goal - this variable could just be helpful to bail out of a function looking specifically for an object while our current goal is a segment. 

I could write this but I want to hear some feedback on if you guys think this may be a good solution. :)

**Comment by vLKp** [comment]
That sounds reasonable.  A possible definition for the structure is:

```c++
struct escort_goal_values {
	escort_goal_t type;
	segnum_t seg;
	objnum_t obj;
};
```

If defined in a cpp file, it should be placed in an anonymous namespace.

**Comment by zicodxx** [comment]
I pushed ad5a5047d4f059c94662ec6844828fbae477a91b which should take care of the messy Escort goal handling. 

First and foremost: It should not cause crashes anymore. Also I cen verify that the reversal of f4b7b958cbc640d1bfd05837f9b041c5c80f0523 fixed the Guidebot behavior in regard to its goals. Keys are now properly detected and I tested for all other goals as well - especially with my fix in place. 

Some details:

After some pondering I decided against the idea of the structure. I still think the variables could use better names but that would have required quite a few changes over several files and I wanted this commit to only address the actual problem.

Reason I decided against the structure: Escort_goal_index did actually not need to be a segment. The only cases in which the variable wasn't pointing to a valid object were: ESCORT_GOAL_SCRAM and ESCORT_GOAL_EXIT. 

ESCORT_GOAL_SCRAM does not need to have Escort_goal_index set at all as it's proposed value of -3 isn't queried anywhere (nor is it actually needed due to special treatment of ESCORT_GOAL_SCRAM, see below). 
ESCORT_GOAL_EXIT likewise does not need Escort_goal_index either. I changed the handling of this goal to work just like ESCORT_GOAL_ENERGYCEN and since this goal is technically never accomplished, it does not need a detection on if the goal is reached. 
For all these goals, I made sure that Escort_goal_index is set to object_none.

Finally, detect_escort_goal_accomplished() was the culprit for the crashes. This function is specifically intended for object goals, only. This function now bails if Escort_goal_index is set to object_none. This replaces the special case for ESCORT_GOAL_SCRAM which now uses object_none, too. 

With this, Escort_goal_index is strictly and only used for objects. This also makes changes regarding this variable in ai_save/restore_state() unnecessary.
